include/felspar/memory/bitmap.strategy.hpp

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+
+#include <cstddef>
+
+
+namespace felspar::memory::bitmap {
+
+
+    /**
+     * The bitmap allocator uses a bitmap to store which blocks are free and
+     * which are currently in use. A set bit indicates a used location and a
+     * zero bit indicates allocated. All memory requests must be for less than
+     * the block size.
+     */
+
+
+    /// The number of bits available in the type
+    template<typename BM>
+    constexpr std::size_t bitcount = sizeof(BM) * 8;
+
+
+    /// Return the memory location that is allocated
+    template<typename BM>
+    constexpr inline std::size_t nextbit(BM allocations) {
+        std::size_t bit{};
+        for (BM mask{1}; bit < bitcount<BM> and allocations & mask;
+             ++bit, mask = (mask << 1u)) {}
+        return bit;
+    }
+
+
+    template<typename BM>
+    std::byte *allocate(BM &bm, std::byte *base, std::size_t blocksize) {
+        if (auto bit = nextbit(bm); bit < bitcount<BM>) {
+            bm |= 1 << bit;
+            return base + blocksize * bit;
```

Comment:
I think here we missing logic for blocksize > 1

case 1
assume that `BM` is type `byte` and current binary of `bm` is 1111'1110 (do 8 allocate then do 1 deallocate)
If we do allocate with blocksize = 2 should it return nullptr?

case 2
assume that `BM` is type `byte` and current binary of `bm` is 0000'0000 (never allocate)
If we do allocate with blocksize = 2 it return `base`
but if current binary of `bm` is 0000'0001 it will return `base + 2`
when we allocate with blocksize = 2 what location caller must use 
return_location and return_location+1 or 
return_location and return_location-1

Reply:
You're exactly right. That's another more complicated version of this allocator. For the time being it's important that the block size is big enough for the biggest expected allocation.

The main use case for a bitmap allocator is for allocating blocks of memory which are all the same size, for example nodes in a `std::map`

Reply:
So it fine for now

Reply:
The complication is that you can't tell what block size was issued purely from the pointer, so you need some extra storage somewhere that tells you that. This complicates things a lot, and I want the allocator to be as fast as possible when it doesn't need to worry about those details